syntax = "proto3";

package benchmark;

option go_package = "./pb";

// BenchmarkService wraps the Python guest module for cross-language benchmarking.
// Each RPC maps to one of the 7 benchmark scenarios.
service BenchmarkService {
  // Scenario 1: void call (wait_a_bit(0))
  rpc VoidCall(VoidCallRequest) returns (VoidCallResponse);

  // Scenario 2: primitive echo (div_integers(10, 2) -> 5.0)
  rpc DivIntegers(DivIntegersRequest) returns (DivIntegersResponse);

  // Scenario 3: string echo (join_strings(["hello","world"]) -> "hello,world")
  rpc JoinStrings(JoinStringsRequest) returns (JoinStringsResponse);

  // Scenario 4: array sum (accepts_ragged_array([[1..N]]) -> sum)
  rpc ArraySum(ArraySumRequest) returns (ArraySumResponse);

  // Scenario 5: object method (SomeClass("bench").print())
  rpc ObjectMethod(ObjectMethodRequest) returns (ObjectMethodResponse);

  // Scenario 6: callback via bidirectional streaming
  // Flow: client sends Invoke -> server sends Compute(a,b) -> client sends Result(sum) -> server sends Final(sum)
  rpc CallbackAdd(stream CallbackClientMsg) returns (stream CallbackServerMsg);

  // Scenario 7: error propagation (returns_an_error() -> gRPC INTERNAL error)
  rpc ReturnsAnError(Empty) returns (Empty);
}

// --- Scenario 1 ---
message VoidCallRequest {
  int64 secs = 1;
}
message VoidCallResponse {}

// --- Scenario 2 ---
message DivIntegersRequest {
  int64 x = 1;
  int64 y = 2;
}
message DivIntegersResponse {
  double result = 1;
}

// --- Scenario 3 ---
message JoinStringsRequest {
  repeated string values = 1;
}
message JoinStringsResponse {
  string result = 1;
}

// --- Scenario 4 ---
message ArraySumRequest {
  repeated int64 values = 1;
}
message ArraySumResponse {
  int64 sum = 1;
}

// --- Scenario 5 ---
message ObjectMethodRequest {
  string name = 1;
}
message ObjectMethodResponse {
  string result = 1;
}

// --- Scenario 6 (callback via bidirectional streaming) ---
message CallbackClientMsg {
  oneof msg {
    bool invoke = 1;        // "please call call_callback_add"
    int64 add_result = 2;   // "result of add(a, b)"
  }
}
message CallbackServerMsg {
  oneof msg {
    CallbackArgs compute = 1;  // "please compute add(a, b)"
    int64 final_result = 2;    // "here's the final result"
  }
}
message CallbackArgs {
  int64 a = 1;
  int64 b = 2;
}

// --- Scenario 7 ---
message Empty {}
